x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportCreateEnrollmentTokenAction.java

```diff
@@ -0,0 +1,147 @@
+            environment, nodeService, sslService, new ApiKeyGenerator(apiKeyService, rolesStore, xContentRegistry));
+    }
+
+    // Constructor for testing
+    TransportCreateEnrollmentTokenAction(TransportService transportService, ActionFilters actionFilters, SecurityContext context,
+                                         Environment environment, NodeService nodeService, SSLService sslService,
+                                         ApiKeyGenerator generator) {
+        super(CreateEnrollmentTokenAction.NAME, transportService, actionFilters, CreateEnrollmentTokenRequest::new);
+        this.generator = generator;
+        this.securityContext = context;
+        this.environment = environment;
+        this.nodeService = nodeService;
+        this.sslService = sslService;
+    }
+
+    @Override
+    protected void doExecute(Task task, CreateEnrollmentTokenRequest request,
+                             ActionListener<CreateEnrollmentTokenResponse> listener) {
+        try {
+            final KeyConfig keyConfig = sslService.getHttpTransportSSLConfiguration().keyConfig();
+            if (keyConfig instanceof StoreKeyConfig == false) {
+                listener.onFailure(new IllegalStateException(
+                    "Unable to create an enrollment token. Elasticsearch node HTTP layer SSL configuration is not configured with a " +
+                        "keystore"));
+                return;
+            }
+            final List<Tuple<PrivateKey, X509Certificate>> httpCaKeysAndCertificates =
+                ((StoreKeyConfig) keyConfig).getPrivateKeyEntries(environment).stream()
+                    .filter(t -> t.v2().getBasicConstraints() != -1).collect(Collectors.toList());
+            if (httpCaKeysAndCertificates.isEmpty()) {
+                listener.onFailure(new IllegalStateException(
+                    "Unable to create an enrollment token. Elasticsearch node HTTP layer SSL configuration Keystore doesn't contain any " +
+                        "PrivateKey entries where the associated certificate is a CA certificate"));
+                return;
+            } else if (httpCaKeysAndCertificates.size() > 1) {
+                listener.onFailure(new IllegalStateException(
+                    "Unable to create an enrollment token. Elasticsearch node HTTP layer SSL configuration Keystore contain multiple " +
+                        "PrivateKey entries where the associated certificate is a CA certificate"));
+                return;
+            } else {
+                final TimeValue expiration = TimeValue.timeValueSeconds(ENROLL_API_KEY_EXPIRATION_SEC);
+                final String[] clusterPrivileges = { "enroll" };
+                final List<RoleDescriptor> roleDescriptors = List.of(new RoleDescriptor("create_enrollment_token", clusterPrivileges,
+                    null, null));
+                CreateApiKeyRequest apiRequest = new CreateApiKeyRequest("enrollment_token_API_key_" + UUIDs.base64UUID(),
```

Comment:
> built-in dedicated owner

Since you mentioned origin, do you mean the internal users, e.g. `_xpack` or `_xpack_security`? One issue is that the `access_granted` audit event is by default disabled for these users and consequently their API keys. We could add a new internal user, but I am not sure whether that's justified unless it also be used for other things.

The other issue is that the extra level of indrection (`elastic` -> `internal user` -> ApiKey) may add confusion to end users. From the ApiKey's perspective, its origin is also lost because it currently only records its direct owner which is the internal user. Also since this API can potentially be used by users other than just `elastic` (still in discussion?), it could make tracing even more challenging.

Reply:
Thank you for engaging Yang!

Yes, I am thinking of a new origin, like `_xpack_security`. In principle, an origin shouldn't handle many things like the others are doing, there's a perennial backlog task to break them up. A specific origin only for this scope, of generating enrollment tokens, wouldn't be inappropriate from this POV, IMO.

BUT, I now realize I mixed two concepts that don't necessary have to go together:
1. the user that calls the client API (eg `client.admin().cluster()`) or the internal java API (eg `ApiKeyService`).
2. the owner user of the temporary API key

The first point is not really important. It is actually a matter of whether the generate-enrollment-token API fits cleanly with the other internal client APIs, eg the generate API keys API, or if it is better to instead piece the existing client APIs together to obtain the enrollment token.
Upon further thought, I think a new client API for generating enrollment tokens, that uses the java API (as is the case here), is the better option, hence **I retract my proposal to use the internal `client`!**

The second point I propose we discuss in speech. I'll reach out.

Reply:
We got together but haven't explicitly touched on the topic of who owns the short-lived API keys.
After discussions on privileges and client enrollment, I think it's fine, at this stage, that the owner be the caller.